tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/PartitionInfo.java

```diff
@@ -0,0 +1,424 @@
+// © 2017 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+package org.unicode.icu.tool.cldrtoicu.localedistance;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import com.google.common.base.CharMatcher;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import com.google.common.collect.SortedSetMultimap;
+import com.google.common.collect.TreeMultimap;
+import com.ibm.icu.impl.locale.LSR;
+
+/**
+ * Provides mapping arrays to quickly lookup partition information for any region
+ * code in client libraries.
+ *
+ * <p>A region's partition is defined by the set of region variables (e.g. "$enUS")
+ * in the CLDR data. Each unique combination of variables forms a partition, and
+ * groups of partitions uniquely define language distance groupings. In slightly
+ * mathematical terms, partition groups form an "equivalence class" for regions
+ * with respect to language distance.
+ *
+ * <p>So by determining the minimum set of partitions and partition groups, and
+ * assigning short IDs to them, it's possibe to create data structures which
+ * support all region pairings while being small and fast to access in client code.
+ */
+final class PartitionInfo {
```

Comment:
Equivalent to:
https://github.com/unicode-org/icu/blob/master/icu4j/tools/misc/src/com/ibm/icu/dev/tool/locale/LocaleDistanceBuilder.java#L652

* Renamed since it doesn't actually "map regions" (it maps variables to partitions).
* Split into a builder and a final value to help encapsulate logic and internal data structures.